tools/src/main/java/org/apache/kafka/tools/VerifiableProducer.java

```diff
@@ -319,59 +392,69 @@ public void onCompletion(RecordMetadata recordMetadata, Exception e) {
+    public void run(ThroughputThrottler throttler) {
 
-                // Flush any remaining messages
-                producer.close();
+        printJson(new StartupComplete());
+        // negative maxMessages (-1) means "infinite"
+        long maxMessages = (this.maxMessages < 0) ? Long.MAX_VALUE : this.maxMessages;
 
-                // Print a summary
-                long stopMs = System.currentTimeMillis();
-                double avgThroughput = 1000 * ((producer.numAcked) / (double) (stopMs - startMs));
-
-                Map<String, Object> data = new HashMap<>();
-                data.put("name", "tool_data");
-                data.put("sent", producer.numSent);
-                data.put("acked", producer.numAcked);
-                data.put("target_throughput", producer.throughput);
-                data.put("avg_throughput", avgThroughput);
-
-                System.out.println(producer.toJsonString(data));
-            }
-        });
-
-        ThroughputThrottler throttler = new ThroughputThrottler(producer.throughput, startMs);
-        System.out.println(producer.startupString());
-        long maxMessages = infinite ? Long.MAX_VALUE : producer.maxMessages;
         for (long i = 0; i < maxMessages; i++) {
-            if (producer.stopProducing) {
+            if (this.stopProducing) {
                 break;
             }
             long sendStartMs = System.currentTimeMillis();
 
-            producer.send(null, producer.getValue(i));
+            this.send(null, this.getValue(i));
 
             if (throttler.shouldThrottle(i, sendStartMs)) {
                 throttler.throttle();
             }
         }
     }
 
+    public static void main(String[] args) {
+        ArgumentParser parser = argParser();
+        if (args.length == 0) {
+            parser.printHelp();
+            Exit.exit(0);
+        }
+
+        try {
+            final VerifiableProducer producer = createFromArgs(parser, args);
+
+            final long startMs = System.currentTimeMillis();
+            ThroughputThrottler throttler = new ThroughputThrottler(producer.throughput, startMs);
+
+            Runtime.getRuntime().addShutdownHook(new Thread() {
+                @Override
+                public void run() {
+                    // Trigger main thread to stop producing messages
+                    producer.stopProducing = true;
+
+                    // Flush any remaining messages
+                    producer.close();
+
+                    // Print a summary
+                    long stopMs = System.currentTimeMillis();
+                    double avgThroughput = 1000 * ((producer.numAcked) / (double) (stopMs - startMs));
+
+                    producer.printJson(new ToolData(producer.numSent, producer.numAcked, producer.throughput, avgThroughput));
+                }
+            });
+
+            producer.run(throttler);
+        } catch (ArgumentParserException e) {
+            parser.handleError(e);
```

Comment:
It seems like we've changed the behavior where no arguments just produces the normal help screen? Or is `handleError` already doing that?